src/Stack/Fetch.hs

```diff
@@ -274,38 +276,66 @@ withCabalLoader menv inner = do
-                Left e -> join $ modifyMVar updateRef $ \toUpdate ->
-                    if toUpdate
-                        then do
+                Nothing -> do
+                    let fuzzy = fuzzyLookupCandidates ident cachesCurr
+                        fuzzyCandidatesText = case fuzzy of
+                            Nothing -> ""
+                            Just cs -> "Possible candidates: "
+                                       <> commaSeparatedIdents cs
+                                       <> ". "
+                    join $ modifyMVar updateRef $ \toUpdate ->
+                        if toUpdate then do
                             runInBase $ do
                                 $logInfo $ T.concat
                                     [ "Didn't see "
                                     , T.pack $ packageIdentifierString ident
-                                    , " in your package indices, updating and trying again"
+                                    , " in your package indices. "
+                                    , T.pack fuzzyCandidatesText
+                                    , "Updating and trying again."
                                     ]
                                 updateAllIndices menv
                                 caches <- getPackageCaches menv
                                 liftIO $ writeIORef icaches caches
                             return (False, doLookup ident)
-                        else return (toUpdate, throwM e)
-
-        doLookup' ident = do
-            caches <- liftIO $ readIORef icaches
-            case Map.lookup ident caches of
-                Nothing ->
-                    return $ Left $ UnknownPackageIdentifiers $ Set.singleton ident
-                Just (index, cache) -> do
-                    [bs] <- flip runReaderT env
-                          $ withCabalFiles (indexName index) [(ident, cache, ())]
-                          $ \_ _ bs -> return bs
-                    return $ Right bs
-
+                        else return (toUpdate, throwM (unknownIdent ident))
     inner doLookup
+  where
+    unknownIdent = UnknownPackageIdentifiers . Set.singleton
+    commaSeparatedIdents = F.fold . intersperse ", " . map packageIdentifierString
+
+type PackageCaches = Map PackageIdentifier (PackageIndex, PackageCache)
+
+lookupPackageIdentifierExact :: HasConfig env
+                             => PackageIdentifier -> env -> PackageCaches
+                             -> IO (Maybe ByteString)
+lookupPackageIdentifierExact ident env caches = do
+    case Map.lookup ident caches of
+        Nothing -> return Nothing
+        Just (index, cache) -> do
+            [bs] <- flip runReaderT env
+                  $ withCabalFiles (indexName index) [(ident, cache, ())]
+                  $ \_ _ bs -> return bs
+            return $ Just bs
+
+-- TODO: use 'Maybe (NonEmpty PackageIdentifier)' return-type
+fuzzyLookupCandidates :: PackageIdentifier -> PackageCaches
```

Comment:
@snoyberg yeah, I decided to not go for efficiency, as this seems like a rare piece of functionality. But since you mentioned it – I have no problem making it efficient then.